### PR TITLE
`tsh proxy kube` should support Machine ID credentials

### LINKED ISSUE #36833: ERROR: access denied: identity is not allowed to reissue certificates while using Machine ID, Layer 7 proxy and "tbot kube proxy"
I am having trouble running Kubernetes proxy with `tbot kube proxy` while using Machine ID credentials. 

The error message I receive is `ERROR: access denied: identity is not allowed to reissue certificates`. 

My Teleport instance is running on an ECS Fargate service over a Layer 7 proxy (AWS ALB), and there is an „agent” node on another AWS Account. However, everything works fine when I use my personal credentials. Its very strange, because im following manual steps, when creating Machine ID credentials.

Teleport version: 14.3.3

**Connecting to cluster using user credentials (works very well)**
```shell
$ tsh login

$ tsh kube ls 
Kube Cluster Name         Labels     Selected
------------------------- ---------- --------
my-kubernetes-eks-cluster ...        *

$ tsh kube proxy --port=4321 my-kubernetes-eks-cluster

$ export KUBECONFIG=/Users/(...)/localproxy-4321-kubeconfig

$ kubectl version
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.5-eks-5e0fdde
```

**Connecting to cluster, using Machine ID credentials (failing)**
```shell
$ tbot start --config=tbot.yaml
INFO [TBOT]      Created directory "/var/lib/teleport/bot" config/destination_directory.go:126
INFO [TBOT]      Created directory "/opt/machine-id" config/destination_directory.go:126
WARN [TBOT]      Failed to open file securely due to missing syscall; falling back to regular file handling. Configure `symlinks: insecure` for "/var/lib/teleport/bot/.write-test" to disable this warning. botfs/fs_linux.go:101
INFO [TBOT]      Initializing bot identity. tbot/tbot.go:254
INFO [TBOT]      Fetching bot identity using token. tbot/bot_identity.go:192
INFO [AUTH]      Attempting registration via proxy server. auth/register.go:285
INFO [AUTH]      Attempting to register Bot with IAM method using regional STS endpoint auth/register.go:663
INFO [AUTH]      Successfully registered Bot with IAM method using regional STS endpoint auth/register.go:696
INFO [AUTH]      Successfully registered via proxy server. auth/register.go:292
INFO [TBOT]      Fetched new bot identity. identity:valid: after=2024-01-17T20:53:17Z, before=2024-01-17T21:54:17Z, duration=1h1m0s | kind=tls, renewable=false, disallow-reissue=false, roles=[bot-terraform], principals=[-teleport-internal-join], generation=0 tbot/tbot.go:298
INFO [TBOT]      Bot initialization complete. tbot/tbot.go:316
INFO [TBOT]      One-shot mode enabled. Generating outputs. tbot/tbot.go:118
INFO [TBOT]      Generating output. output:kubernetes (directory: /opt/machine-id) tbot/impersonated_identity.go:550
INFO [TBOT]      Generated identity for Kubernetes cluster "my-kubernetes-eks-cluster" tbot/impersonated_identity.go:478
INFO [TBOT]      Generated output. output:kubernetes (directory: /opt/machine-id) tbot/impersonated_identity.go:595
INFO [TBOT]      Generated outputs. One-shot mode is enabled so exiting. tbot/tbot.go:123

$ tsh proxy --proxy=proxy.teleport-host.cloud:443 kube --identity=/opt/machine-id/identity --port=4321 my-kubernetes-eks-cluster
Preparing the following Teleport Kubernetes clusters:
Teleport Cluster Name Kube Cluster Name
--------------------- -------------------------------------
my-teleport-cluster   my-kubernetes-eks-cluster

ERROR: access denied: identity is not allowed to reissue certificates

$ tbot proxy --config=tbot.yaml --proxy=proxy.teleport-host.cloud:443 kube my-kubernetes-eks-cluster
2024-01-17T20:59:40Z [CLIENT]    INFO ALPN connection upgrade required for "proxy.teleport-host.cloud:443": false. client/api.go:728
2024-01-17T20:59:40Z [CLIENT]    INFO no host login given. defaulting to root client/api.go:1066
2024-01-17T20:59:40Z [CLIENT]    DEBU Skipping connection to the local ssh-agent. client/keyagent.go:137
2024-01-17T20:59:40Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:40Z [KEYAGENT]  INFO Loading SSH key for user "bot-terraform" and cluster "my-teleport-cluster". client/keyagent.go:196
2024-01-17T20:59:40Z [TSH]       DEBU Pinging the proxy to fetch listening addresses for non-web ports. common/tsh.go:3452
2024-01-17T20:59:40Z [CLIENT]    DEBU not using loopback pool for remote proxy addr: proxy.teleport-host.cloud:443 client/api.go:4527
2024-01-17T20:59:40Z             DEBU Attempting GET proxy.teleport-host.cloud:443/webapi/ping webclient/webclient.go:129
2024-01-17T20:59:40Z             DEBU ALPN connection upgrade required for "proxy.teleport-host.cloud:443": true. client/alpn_conn_upgrade.go:95
2024-01-17T20:59:40Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:40Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:40Z [CLIENT]    INFO Connecting to proxy=proxy.teleport-host.cloud:443 login="-teleport-nologin-2d420e0f-bd4e-4207-ac0c-0c10ce5c33c8" using TLS Routing client/api.go:3055
2024-01-17T20:59:40Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:40Z [HTTP:PROX] DEBU No proxy set in environment, returning direct dialer. proxy/proxy.go:195
2024-01-17T20:59:40Z             DEBU ALPN connection upgrade for proxy.teleport-host.cloud:443. client/alpn_conn_upgrade.go:187
2024-01-17T20:59:41Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:41Z [KEYAGENT]  DEBU "Checking key: ssh-rsa-cert-v01@openssh.com (redacted)\n." client/keyagent.go:368
2024-01-17T20:59:41Z [KEYAGENT]  DEBU Validated host proxy.teleport-host.cloud:443. client/keyagent.go:374
2024-01-17T20:59:41Z [CLIENT]    INFO Successful auth with proxy proxy.teleport-host.cloud:443. client/api.go:3060
2024-01-17T20:59:41Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:41Z             DEBU ALPN connection upgrade for proxy.teleport-host.cloud:443. client/alpn_conn_upgrade.go:187
Preparing the following Teleport Kubernetes clusters:
Teleport Cluster Name Kube Cluster Name
--------------------- -------------------------------------
my-teleport-cluster           my-kubernetes-eks-cluster

2024-01-17T20:59:41Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:41Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:41Z [CLIENT]    INFO Connecting to proxy=proxy.teleport-host.cloud:443 login="-teleport-nologin-2d420e0f-bd4e-4207-ac0c-0c10ce5c33c8" using TLS Routing client/api.go:3055
2024-01-17T20:59:41Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:41Z [HTTP:PROX] DEBU No proxy set in environment, returning direct dialer. proxy/proxy.go:195
2024-01-17T20:59:41Z             DEBU ALPN connection upgrade for proxy.teleport-host.cloud:443. client/alpn_conn_upgrade.go:187
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:42Z [KEYAGENT]  DEBU "Checking key: ssh-rsa-cert-v01@openssh.com (redacted)\n." client/keyagent.go:368
2024-01-17T20:59:42Z [KEYAGENT]  DEBU Validated host proxy.teleport-host.cloud:443. client/keyagent.go:374
2024-01-17T20:59:42Z [CLIENT]    INFO Successful auth with proxy proxy.teleport-host.cloud:443. client/api.go:3060
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:42Z             DEBU ALPN connection upgrade for proxy.teleport-host.cloud:443. client/alpn_conn_upgrade.go:187
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:42Z [CLIENT]    DEBU MFA not required for access. client/client.go:508
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106
2024-01-17T20:59:42Z [KEYSTORE]  DEBU Teleport TLS certificate valid until "2024-01-17 21:54:22 +0000 UTC". client/client_store.go:106

ERROR REPORT:
Original Error: *trace.AccessDeniedError access denied: identity is not allowed to reissue certificates
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/client/client.go:949 github.com/gravitational/teleport/api/client.(*Client).GenerateUserCerts
	github.com/gravitational/teleport/lib/client/client.go:363 github.com/gravitational/teleport/lib/client.(*ProxyClient).reissueUserCerts
	github.com/gravitational/teleport/lib/client/client.go:516 github.com/gravitational/teleport/lib/client.(*ProxyClient).IssueUserCertsWithMFA
	github.com/gravitational/teleport/tool/tsh/common/kube_proxy.go:571 github.com/gravitational/teleport/tool/tsh/common.issueKubeCert
	github.com/gravitational/teleport/tool/tsh/common/kube_proxy.go:488 github.com/gravitational/teleport/tool/tsh/common.loadKubeUserCerts
	github.com/gravitational/teleport/tool/tsh/common/kube_proxy.go:295 github.com/gravitational/teleport/tool/tsh/common.makeKubeLocalProxy
	github.com/gravitational/teleport/tool/tsh/common/kube_proxy.go:112 github.com/gravitational/teleport/tool/tsh/common.(*proxyKubeCommand).run
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:1354 github.com/gravitational/teleport/tool/tsh/common.Run
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:548 github.com/gravitational/teleport/tool/tsh/common.Main
	github.com/gravitational/teleport/tool/tsh/main.go:24 main.main
	runtime/proc.go:267 runtime.main
	runtime/asm_amd64.s:1650 runtime.goexit
User Message: access denied: identity is not allowed to reissue certificates
ERROR: executing `tsh proxy`
	unable to execute tsh
		exit status 1
```

Bot user "terraform", and role "bot-terraform" created using:
```shell
$ tctl bots add terraform --roles=terraform --token=terraform
```

_token_:
```yaml
kind: token
metadata:
  name: terraform
spec:
  allow:
  - aws_account: "(redacted)"
  - aws_account: "(redacted)"
  - aws_account: "(redacted)"
  bot_name: terraform
  join_method: iam
  roles:
  - Bot
version: v2
```

_tbot.yaml_
```yaml
version: v2
debug: true
onboarding:
  join_method: iam
  token: terraform
storage:
  type: directory
  path: /var/lib/teleport/bot
outputs:
- type: kubernetes
  kubernetes_cluster: my-kubernetes-eks-cluster
  destination:
    type: directory
    path: "/opt/machine-id"
auth_server: proxy.teleport-host.cloud:443
certificate_ttl: 1h0m0s
renewal_interval: 20m
oneshot: true
fips: false
```

_bot-terraform role_:
```yaml
kind: role
metadata:
  description: Automatically generated role for bot terraform
  id: 1705521879055712877
  labels:
    teleport.internal/bot: terraform
  name: bot-terraform
spec:
  allow:
    impersonate:
      roles:
      - terraform
    rules:
    - resources:
      - cert_authority
      verbs:
      - readnosecrets
  deny: {}
  options:
    cert_format: standard
    create_db_user: false
    create_desktop_user: false
    desktop_clipboard: true
    desktop_directory_sharing: true
    enhanced_recording:
    - command
    - network
    forward_agent: false
    idp:
      saml:
        enabled: true
    max_session_ttl: 12h0m0s
    pin_source_ip: false
    port_forwarding: true
    record_session:
      default: best_effort
      desktop: true
    ssh_file_copy: true
version: v7
```

_terraform role_
```yaml
kind: role
metadata:
  id: 1705519614671541101
  name: terraform
spec:
  allow:
    app_labels:
      '*': '*'
    db_labels:
      '*': '*'
    db_names:
    - '*'
    db_users:
    - '*'
    kubernetes_groups:
    - '*'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
    - kind: '*'
      name: '*'
      namespace: '*'
    rules:
    - resources:
      - user
      - role
      - token
      - trusted_cluster
      - github
      - oidc
      - saml
      - cluster_auth_preference
      - cluster_networking_config
      - cluster_maintenance_config
      - session_recording_config
      - app
      - db
      - kube_cluster
      - login_rule
      - device
      - okta_import_rule
      - access_list
      verbs:
      - list
      - create
      - read
      - update
      - delete
  deny: {}
  options:
    cert_format: standard
    create_db_user: false
    create_desktop_user: false
    desktop_clipboard: true
    desktop_directory_sharing: true
    enhanced_recording:
    - command
    - network
    forward_agent: false
    idp:
      saml:
        enabled: true
    max_session_ttl: 30h0m0s
    pin_source_ip: false
    port_forwarding: true
    record_session:
      default: best_effort
      desktop: true
    ssh_file_copy: true
version: v7
```

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/36833

changelog: Adds `tbot proxy kube` to support connecting to Kubernetes clusters using Machine ID when the Proxy is behind a L7 LB.

### CODE REVIEW

**Comment by strideynet** [comment]
CC @Joerger - this is the change I ran past you yesterday in case you had any further comments.

**Review by Joerger — APPROVED** [review]

File: tool/tbot/proxy.go [hunk 1]
```diff
@@ -63,5 +65,20 @@ func onProxyCommand(botConfig *config.BotConfig, cf *config.CLIConf) error {
...
+		env[teleport.EnvKubeConfig] = filepath.Join(
+			destination.Path, "kubeconfig.yaml",
+		)
+		// `tsh kube proxy` uses TELEPORT_KUBECONFIG to determine where to write
+		// the modified kube config file intended for proxying.
+		env["TELEPORT_KUBECONFIG"] = filepath.Join(
+			destination.Path, "kubeconfig-proxied.yaml",
+		)
+	}
+
```
Comment (Tener) [inline tool/tbot/proxy.go:82]: Nit: I wonder if we could have a test for this behaviour.

**Review by Tener — APPROVED** [review]

**Review by ravicious — APPROVED** [review]

**Comment by Tener** [comment]
@strideynet 

https://github.com/gravitational/teleport/actions/runs/7624334133/job/20766267419

```
=== Failed
=== FAIL: integration TestIntegrations/Disconnection/expired_cert_node_recording (7.07s)

Error copying data to remote peer
goroutine 47576 [running]:
```

**Comment by strideynet** [comment]
@Tener 

Looks like https://github.com/gravitational/teleport/issues/36085

**Pull request merged**